clients/application/src/main/java/org/eclipse/hono/application/client/CommandSender.java

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.hono.application.client;
+
+import java.util.Map;
+
+import org.eclipse.hono.client.ServiceInvocationException;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * A client for sending commands.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/command-and-control/">
+ *      Command &amp; Control API for AMQP 1.0 Specification</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/command-and-control-kafka/">
+ *      Command &amp; Control API for Kafka Specification</a>
+ */
+public interface CommandSender {
+    /**
+     * Sends an async command to a device, i.e. there is no immediate response expected from the device, but
+     * asynchronously via a separate consumer.
+     * <p>
+     * A device needs to be (successfully) registered before a client can upload any data for it. The device also needs
+     * to be connected for a successful delivery.
+     *
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device to send the command to.
+     * @param command The command name.
+     * @param data The command data to send to the device or {@code null} if the command has no input data.
+     * @param correlationId The identifier to use for correlating the response with the request. Note: This parameter is
+     *            security sensitive. To ensure secure request response mapping choose correlationId carefully, e.g.
+     *            {@link java.util.UUID#randomUUID()}.
+     * @param replyId This identifier is to be set depending upon the underlying implementation (i.e AMQP or Kafka based).
+     *                <ul>
+     *                <li>In Kafka based implementation, this is not used and hence it can be set as {@code null}.</li>
+     *                <li>In AMQP based implementation, this is an arbitrary string which gets used for the response 
+     *                link address in the form of <em>command_response/${tenantId}/${replyId}</em>. Must match the
+     *                {@code replyId} passed to the command response receiver. The replyId must not be {@code null}</li>
+     *                </ul>
```

Comment:
It feels very strange to refer to a particular (AMQP based) implementation here when the whole point of the interface is to abstract away the underlying transport protocol. If we refer to a corresponding *command response receiver* then it also feels strange that there is no method/way of getting one. Based on the fact that the replyId seems ot be relevant for the AMQP based implementation, I wonder why that is. If in fact we do not need the replyId at all, then why not simply remove it from here completely?

Reply:
The `replyId` is needed for AMQP because we support command response addresses with the custom `replyId` suffix there (`command_response/${tenantId}/${replyId}`). 
For the Kafka implementation, there is only the fixed `hono.command_response.${tenantId}` topic on which command responses will be sent and received (thereby not increasing the overall number of topics unnecessarily). 

Of course we _could_ get rid of the `replyId` in the AMQP implementation but that would be a bigger change (would also be an API break) and would be something to discuss separately.

For now, I think the javadoc could also be rephrased to keep it more generic: E.g. something like "The `replyId` is used as part of the resource to which a corresponding command response should be sent, if the messaging network specific implementation defines the use of such a variable part of the endpoint identifier. If not, `null` is to be used here."

Reply:
IMHO we should actually describe it the other way around, i.e.
> If the underlying messaging infrastructure does not require a replyId, the specified value will be ignored.

WDYT?

Reply:
Since it's not a requirement of the messaging infrastructure but rather of the Command & Control implementation on top of it, I would prefer:

> If the messaging network specific Command & Control implementation does not require a replyId, the specified value will be ignored.

Reply:
It sounds good. Just to be sure, I would add to the JavaDoc as below.

```
     * @param replyId An arbitrary string which will be used to create the reply-to address to be included in
     *                commands sent to devices of the tenant. If the messaging network specific Command & Control
     *                implementation does not require a replyId, the specified value will be ignored.
```

Reply:
@sophokles73 I have updated it and pushed the changes. Would you mind taking a look?